File: goodjava/src/goodjava/webserver/RequestParser.java
package goodjava.webserver;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.util.List;
import java.util.ArrayList;
import goodjava.logging.Logger;
import goodjava.logging.LoggerFactory;
import goodjava.parser.Parser;
import goodjava.parser.ParseException;


final class RequestParser {
	private static final Logger logger = LoggerFactory.getLogger(RequestParser.class);
	private final Request request;
	private Parser parser;

	RequestParser(Request request) {
		this.request = request;
	}

	void parseUrlencoded(String charset) throws ParseException, UnsupportedEncodingException {
		if( request.body == null ) {
			logger.warn("body is null\n"+request.rawHead);
			return;
		}
		this.parser = new Parser(Util.toString(request.body,charset));
		parseQuery();
		require( parser.endOfInput() );
	}

	void parseHead() throws ParseException {
		this.parser = new Parser(request.rawHead);
		parseRequestLine();
		while( !parser.match("\r\n") ) {
			parserHeaderField();
		}
		parseCookies();
	}

	private void parseRequestLine() throws ParseException {
		parseMethod();
		require( parser.match(' ') );
		parseRawPath();
		require( parser.match(' ') );
		parseProtocol();
		require( parser.match("\r\n") );
	}

	private void parseMethod() throws ParseException {
		int start = parser.currentIndex();
		if( !methodChar() )
			throw new ParseException(parser,"no method");
		while( methodChar() );
		request.method = parser.textFrom(start);
	}

	private boolean methodChar() {
		return parser.inCharRange('A','Z');
	}

	private void parseRawPath() throws ParseException {
		int start = parser.currentIndex();
		parsePath();
		if( parser.match('?') )
			parseQuery();
		request.rawPath = parser.textFrom(start);
	}

	private void parsePath() throws ParseException {
		int start = parser.currentIndex();
		if( !parser.match('/') )
			throw new ParseException(parser,"bad path");
		while( parser.noneOf(" ?#") );
		request.path = urlDecode( parser.textFrom(start) );
		request.originalPath = request.path;
	}

	private void parseQuery() throws ParseException {
		do {
			int start = parser.currentIndex();
			while( queryChar() );
			String name = urlDecode( parser.textFrom(start) );
			String value = null;
			if( parser.match('=') ) {
				start = parser.currentIndex();
				while( queryChar() || parser.match('=') );
				value = urlDecode( parser.textFrom(start) );
			}
			if( name.length() > 0 || value != null ) {
				if( value==null )
					value = "";
				Util.add(request.parameters,name,value);
			}
		} while( parser.match('&') );
	}

	private boolean queryChar() {
		return parser.noneOf("=&# \t\n\f\r\u000b");
	}

	private void parseProtocol() throws ParseException {
		int start = parser.currentIndex();
		if( !(
			parser.match("HTTP/")
			&& parser.inCharRange('0','9')
			&& parser.match('.')
			&& parser.inCharRange('0','9')
		) )
			throw new ParseException(parser,"bad protocol");
		request.protocol = parser.textFrom(start);
		request.scheme = "http";
	}


	private void parserHeaderField() throws ParseException {
		String name = parseName();
		require( parser.match(':') );
		while( parser.anyOf(" \t") );
		String value = parseValue();
		while( parser.anyOf(" \t") );
		require( parser.match("\r\n") );
		Util.add(request.headers,name,value);
	}

	private String parseName() throws ParseException {
		int start = parser.currentIndex();
		require( tokenChar() );
		while( tokenChar() );
		return parser.textFrom(start).toLowerCase();
	}

	private String parseValue() throws ParseException {
		int start = parser.currentIndex();
		while( !testEndOfValue() )
			require( parser.anyChar() );
		return parser.textFrom(start);
	}

	private boolean testEndOfValue() {
		parser.begin();
		while( parser.anyOf(" \t") );
		boolean b = parser.endOfInput() || parser.anyOf("\r\n");
		parser.failure();  // rollback
		return b;
	}

	private void require(boolean b) throws ParseException {
		if( !b )
			throw new ParseException(parser,"failed");
	}

	boolean tokenChar() {
		if( parser.endOfInput() )
			return false;
		char c = parser.currentChar();
		if( 32 <= c && c <= 126 && "()<>@,;:\\\"/[]?={} \t\r\n".indexOf(c) == -1 ) {
			parser.anyChar();
			return true;
		} else {
			return false;
		}
	}


	private void parseCookies() throws ParseException {
		String text = (String)request.headers.get("cookie");
		if( text == null )
			return;
		this.parser = new Parser(text);
		while(true) {
			int start = parser.currentIndex();
			while( parser.noneOf("=;") );
			String name = urlDecode( parser.textFrom(start) );
			if( parser.match('=') ) {
				start = parser.currentIndex();
				while( parser.noneOf(";") );
				String value = parser.textFrom(start);
				int len = value.length();
				if( value.charAt(0)=='"' && value.charAt(len-1)=='"' )
					value = value.substring(1,len-1);
				value = urlDecode(value);
				request.cookies.put(name,value);
			}
			if( parser.endOfInput() )
				return;
			require( parser.match(';') );
			parser.match(' ');  // optional for bad browsers
		}
	}


	private static final String contentTypeStart = "multipart/form-data; boundary=";

	void parseMultipart() throws ParseException, UnsupportedEncodingException {
		if( request.body == null ) {
			logger.warn("body is null\n"+request.rawHead);
			return;
		}
		String contentType = (String)request.headers.get("content-type");
		if( !contentType.startsWith(contentTypeStart) )
			throw new RuntimeException(contentType);
		String boundary = "--"+contentType.substring(contentTypeStart.length());
		this.parser = new Parser(Util.toString(request.body,null));
//System.out.println(this.parser.text);
		require( parser.match(boundary) );
		boundary = "\r\n" + boundary;
		while( !parser.match("--\r\n") ) {
			require( parser.match("\r\n") );
			require( parser.match("Content-Disposition: form-data; name=") );
			String name = quotedString();
			String filename = null;
			boolean isBinary = false;
			if( parser.match("; filename=") ) {
				filename = quotedString();
				require( parser.match("\r\n") );
				require( parser.match("Content-Type: ") );
				int start = parser.currentIndex();
				if( parser.match("application/") ) {
					isBinary = true;
				} else if( parser.match("image/") ) {
					isBinary = true;
				} else if( parser.match("text/") ) {
					isBinary = false;
				} else
					throw new ParseException(parser,"bad file content-type");
				while( parser.inCharRange('a','z') || parser.anyOf("-.") );
				contentType = parser.textFrom(start);
			}
			require( parser.match("\r\n") );
			require( parser.match("\r\n") );
			int start = parser.currentIndex();
			while( !parser.test(boundary) ) {
				require( parser.anyChar() );
			}
			String value = parser.textFrom(start);
			if( filename == null ) {
				Util.add(request.parameters,name,value);
			} else {
				Object content = isBinary ? Util.toBytes(value) : value;
				Request.MultipartFile mf = new Request.MultipartFile(filename,contentType,content);
				Util.add(request.parameters,name,mf);
			}
			require( parser.match(boundary) );
		}
	}

	private String quotedString() throws ParseException {
		StringBuilder sb = new StringBuilder();
		require( parser.match('"') );
		while( !parser.match('"') ) {
			if( parser.match("\\\"") ) {
				sb.append('"');
			} else {
				require( parser.anyChar() );
				sb.append( parser.lastChar() );
			}
		}
		return sb.toString();
	}

	private String urlDecode(String s) throws ParseException {
		try {
			return URLDecoder.decode(s,"UTF-8");
		} catch(UnsupportedEncodingException e) {
			parser.rollback();
			throw new ParseException(parser,e);
		} catch(IllegalArgumentException e) {
			parser.rollback();
			throw new ParseException(parser,e);
		}
	}

	// improve later
	void parseJson(String charset) throws UnsupportedEncodingException {
		if( request.body == null ) {
			logger.warn("body is null\n"+request.rawHead);
			return;
		}
		String value = Util.toString(request.body,charset);
		Util.add(request.parameters,"json",value);
	}

}

File: webserver/requestparser.py
# from .parser import Parser
# from .request import Request

# class RequestParser:
#     # private static final Logger logger = LoggerFactory.getLogger(RequestParser.class)
#     def __init__(self, request):
#         if isinstance(request,Request):
#             self.request = request
#             self.parser = Parser()
#         else:
#             raise ValueError("Request required")
    
#     def parseurlencoded(self,charset):
#         if self.request.body == None:
#             print(f"body is null\n{self.request.rawhead}")
#             return
#         self.parser = Parser(str(self.request.body,encoding=charset))
#         self.parsequery()
#         self.require(self.parser.endofinput())
    
#     def parsehead(self):
#         self.parser = Parser(self.request.rawhead)
#         self.parserequestline()
#         while not self.parser.match("\r\n"):
#             self.parserheaderfield()
#         self.parsecookies()

#     def parserequestline(self):
#         self.parsemethod()
#         self.require( self.parser.match(' ') )
#         self.parserawpath()
#         self.require( self.parser.match(' ') )
#         self.parseprotocol()
#         self.require( self.parser.match("\r\n") )

#     def parsemethod(self):
#         start = self.parser.currentindex()
#         if self.methodchar():
#             raise ParseException(self.parser,"no method")
#         while self.methodchar():
#             continue
#         self.request.method = self.parser.textfrom(start)

#     def methodchar(self):
#         return self.parser.incharrange('A','Z')

#     def parserawpath(self):
#         start = self.parser.currentindex()
#         self.parsepath()
#         if self.parser.match('?'):
#             self.parsequery()
#         self.request.rawpath = self.parser.textfrom(start)

#     def parsepath(self):
#         start = self.parser.currentindex()
#         if notself.parser.match('/'):
#             raise ParseException(parser,"bad path")
#         while self.parser.noneof(" ?#"):
#             continue
#         self.request.path = self.urldecode( self.parser.textfrom(start) )
#         self.request.originalpath = self.request.path


#     def parsequery(self):
#         while True:
#             start = self.parser.currentindex()
#             while self.querychar():
#                 continue
#             name = self.urldecode(self.parser.textfrom(start))
#             value = None
#             if self.parser.match('='):
#                 start = self.parser.currentindex()
#                 while self.querychar() or self.parser.match('='):
#                     continue
#                 value = self.urldecode(self.parser.textfrom(start))
#             if len(name) > 0 or value != None:
#                 if value == None:
#                     value = ""
#                 self.request.parameters[name] = value
#             if not self.parser.match('&'):
#                 break

#     def querychar(self):
#         return self.parser.noneof("=&# \t\n\f\r\u000b")

#     def parseprotocol(self):
#         start = self.parser.currentindex()
#         if not (self.parser.match("HTTP/") and self.parser.incharrange('0','9') and self.parser.match('.') and self.parser.incharrange('0','9')):
#             raise ParseException(self.parser,"bad protocol")
#         self.request.protocol = self.parser.textfrom(start)
#         self.request.scheme = "http"

#     def parserheaderfield(self):
#         name = self.parsename()
#         self.require(self.parser.match(':'))
#         while self.parser.anyof(" \t"):
#             continue
#         value = self.parsevalue()
#         while self.parser.anyof(" \t"):
#             continue
#         self.require(self.parser.match("\r\n"))
#         self.request.headers[name] = value

#     def parsename(self):
#         start = self.parser.currentindex()
#         self.require(self.tokenchar())
#         while self.tokenchar():
#             continue
#         return self.parser.textfrom(start).lower()

#     def parsevalue(self):
#         start = self.parser.currentindex()
#         while not self.testendofvalue():
#             self.require(self.parser.anychar())
#         return self.parser.textfrom(start)

#     def testendofvalue(self):
#         self.parser.begin()
#         while self.parser.anyof(" \t"):
#             continue
#         isendof = self.parser.endofinput() or self.parser.anyof("\r\n")
#         self.parser.failure() # rollback
#         return isendof

#     def require(self, boolean):
#         if not boolean:
#             raise ParseException(self.parser,"failed")

#     def tokenchar(self):
#         if self.parser.endofinput():
#             return False
#         c = self.parser.currentchar()
#         if 32 <= c <= 126 and "()<>@,:\\\"/[]?={} \t\r\n".indexOf(c) == -1:
#             self.parser.anychar()
#             return True
#         else:
#             return False

#     def parsecookies(self):
#         text = self.request.headers.get("cookie")
#         if text == None:
#             return
#         self.parser = Parser(text)
#         while True:
#             start = self.parser.currentindex()
#             while self.parser.noneof("="):
#                 continue
#             name = self.urldecode(self.parser.textfrom(start))
#             if self.parser.match('='):
#                 start = self.parser.currentindex()
#                 while self.parser.noneof(""):
#                     continue
#                 value = self.parser.textfrom(start)
#                 length = len(value)
#                 if value.charat(0)=='"' and value.charat(length-1) == '"':
#                     value = value.substring(1,length-1)
#                 value = self.urldecode(value)
#                 self.request.cookies[name] = value
#             if self.parser.endofinput():
#                 return
#             self.require(self.parser.match(''))
#             self.parser.match(' ')  # optional for bad browsers

#     def parsemultipart(self):
#         contenttypestart = "multipart/form-data boundary="
#         if self.request.body == None:
#             print("body is null\n"+self.request.rawhead)
#             return
#         contenttype = self.request.headers.get("content-type")
#         if not contenttype.startsWith(contenttypestart):
#             raise RuntimeException(contenttype)
#         boundary = "--" + contenttype.substring(len(contenttypestart))
#         self.parser = Parser(str(self.request.body))
#         #System.out.println(this.parser.text)
#         self.require(self.parser.match(boundary))
#         boundary = "\r\n" + boundary
#         while not self.parser.match("--\r\n"):
#             self.require(self.parser.match("\r\n"))
#             self.require(self.parser.match("Content-Disposition: form-data name="))
#             name = self.quotedstring()
#             filename = None
#             isbinary = False
#             if self.parser.match(" filename="):
#                 filename = self.quotedstring()
#                 self.require(self.parser.match("\r\n"))
#                 self.require(self.parser.match("Content-Type: "))
#                 start = self.parser.currentindex()
#                 if self.parser.match("application/"):
#                     isbinary = True
#                 elif self.parser.match("image/"):
#                     isbinary = True
#                 elif self.parser.match("text/"):
#                     isbinary = False
#                 else:
#                     raise ParseException(self.parser,"bad file content-type")
#                 while self.parser.incharrange('a','z') or self.parser.anyof("-."):
#                     continue
#                 contenttype = self.parser.textfrom(start)
#             self.require(self.parser.match("\r\n"))
#             self.require(self.parser.match("\r\n"))
#             start = self.parser.currentindex()
#             while not self.parser.test(boundary):
#                 self.require(self.parser.anychar())
#             value = self.parser.textfrom(start)
#             if filename == None:
#                 self.request.parameters[name] = value
#             else:
#                 content = bytes(value) if isbinary else value
#                 mf = MultipartFile(filename,contenttype,content)
#                 self.request.parameters[name] = mf
#             self.require(self.parser.match(boundary))

#     def quotedstring(self):
#         sb = ""
#         self.require(self.parser.match('"'))
#         while not self.parser.match('"'):
#             if self.parser.match("\\\""):
#                 sb += '"'
#             else:
#                 self.require(self.parser.anychar())
#                 sb += self.parser.lastchar()
#         return str(sb)

#     def urldecode(self, string):
#         try:
#             return str(string, encoding="utf-8")
#         except UnsupportedEncodingException as uee:
#             self.parser.rollback()
#             raise ParseException(parser,uee)
#         except IllegalArgumentException as iae:
#             self.parser.rollback()
#             raise ParseException(parser,iae)

#     def parsejson(self, charset):
#         if self.request.body == None:
#             print("body is null\n"+self.request.rawhead)
#             return
#         value = str(self.request.body,encoding=charset)
#         self.request.parameters["json"] = value